File: bots/template.js
// Challenge Name Goes Here # TEMPLATE FOR CHROME
const puppeteer = require('puppeteer');
var queue = require('../redis-controller/queue');
const queueName = __filename.split(".")[0].split("/").pop();

const challName = "bot"

const thecookie = {
    "name": 'xoxo',
    "value": "xoxo",
    "domain": 'localhost',
    "httpOnly": true,
    "sameSite": 'Lax',
  }



async function url_visit (url) {
    var quote;
    return new Promise(async function(resolve, reject) {
        // start modification

        const browser = await puppeteer.launch({executablePath: '/home/bot/latest/chrome'}); 
        const page = await browser.newPage();  
        await page.setCookie(thecookie)
        await page.setDefaultNavigationTimeout(1e3*15);  // Timeout duration in milliseconds    // use either this or wait for navigation
        try{
            var result = await page.goto(url);
            await page.waitForNavigation(); // wait till the page finishes loading              
        }
        catch(e){
            console.log();
        }        
        await browser.close();

        // end modification
        resolve(quote);
    });
}



function popMe(){
    queue.pop(queueName,sendUrl)
}

async function sendUrl(err, url) {
    if (err) {
        throw err;
    }

    if (!url) {
        setTimeout(popMe, 1e3); // if null is returned, wait for a sec before retrying
    } else {
        console.log(url)
        q = await url_visit(url)
        popMe();
    } 
}

console.log(`Started bot for chall ${challName} with id ${queueName}`)
popMe()

File: bots/template.py
#  Challenge Name Goes Here // TEMPLATE FOR FIREFOX
import os
import redis
import time
import threading
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from selenium.webdriver import FirefoxOptions


queueName = os.path.basename(__file__).split(".py")[0]

challName = "bot"

r = redis.Redis(host='0.0.0.0', port=6379)

host = "http://localhost"
cookie = {"name": "Token", "value": "1337", "httpOnly": True}

opts = FirefoxOptions()
opts.add_argument("--headless")

def url_visit(url):
    driver = webdriver.Firefox(firefox_options=opts)
    driver.set_page_load_timeout(10)
    try:
        driver.get(host)
        driver.add_cookie(cookie)
        driver.get(url)
        time.sleep(5)
        driver.quit()
        return True
    except:
        print ("Could not load the URL.")
        driver.quit()
        return False


def sendUrl():
    while True:
        try:
            x = popMe()
            print(x)
            url_visit(x)
        except:
            time.sleep(1)
            continue
        

def popMe():
    return (r.rpop(queueName).decode('utf8'))


print(f"Started bot for chall {challName} with id {queueName}")

# sendUrl()
t1 = threading.Thread(target=sendUrl)
t2 = threading.Thread(target=sendUrl)
t3 = threading.Thread(target=sendUrl)

t1.start()
time.sleep(1.1)
t2.start()
time.sleep(2.3)
t3.start()

t1.join()
t2.join()
t3.join()
